controller/controller.go

```diff
@@ -600,22 +600,24 @@ func (ctrl *ProvisionController) HasRun() bool {
 // NewProvisionController creates a new provision controller using
 // the given configuration parameters and with private (non-shared) informers.
 func NewProvisionController(
+	logger klog.Logger,
 	client kubernetes.Interface,
 	provisionerName string,
 	provisioner Provisioner,
 	options ...func(*ProvisionController) error,
 ) *ProvisionController {
 	id, err := os.Hostname()
 	if err != nil {
-		klog.Fatalf("Error getting hostname: %v", err)
+		logger.Error(err, "Error getting hostname")
+		klog.FlushAndExit(klog.ExitFlushTimeout, 1)
 	}
 	// add a uniquifier so that two processes on the same host don't accidentally both become active
 	id = id + "_" + string(uuid.NewUUID())
 	component := provisionerName + "_" + id
 
 	v1.AddToScheme(scheme.Scheme)
 	broadcaster := record.NewBroadcaster()
-	broadcaster.StartLogging(klog.Infof)
+	broadcaster.StartStructuredLogging(0)
```

Comment:
Here's a problem: we need a new API such that `broadcaster` accepts a logger instead of using the global one.

Reply:
Let's not block the PR because of that, but a TODO would be good.

Reply:
You can link to https://github.com/kubernetes/kubernetes/pull/120729.

Reply:
@pohly, I have a couple of questions regarding this:

1. Since no release tag has been created following the merge of the [PR that added NewEventBroadcasterAdapterWithContext](https://github.com/kubernetes/kubernetes/pull/122142), is the adaptation to contextual logging for the event broadcaster unnecessary in this PR?

2. [The interface of ResourceLockConfig.EventRecorder](https://github.com/kubernetes/kubernetes/blob/7811ba3d4d8666f942ffecd57ebd194636e914f7/staging/src/k8s.io/client-go/tools/leaderelection/resourcelock/interface.go#L124-L127) seems different from [the new EventRecorder's interface]((https://github.com/kubernetes/kubernetes/blob/7811ba3d4d8666f942ffecd57ebd194636e914f7/staging/src/k8s.io/client-go/tools/internal/events/interfaces.go#L43)). Would it be okay to change the interface of ResourceLockConfig.EventRecorder in a separate PR?

Reply:
> Since no release tag has been created following the merge of the https://github.com/kubernetes/kubernetes/pull/122142, is the adaptation to contextual logging for the event broadcaster unnecessary in this PR?

Perhaps wait until this repo is updated to Kubernetes 1.30, then use the WithContext call?

> [The interface of ResourceLockConfig.EventRecorder](https://github.com/kubernetes/kubernetes/blob/7811ba3d4d8666f942ffecd57ebd194636e914f7/staging/src/k8s.io/client-go/tools/leaderelection/resourcelock/interface.go#L124-L127) seems different from the new EventRecorder's interface. Would it be okay to change the interface of ResourceLockConfig.EventRecorder in a separate PR?

I would update it together with updating Kubernetes 1.30 and adapting the code.